File: solution/java/week05-complexity/exercise2/src/test/java/command/CommandProcessorTests.java
package command;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;

import java.time.LocalDate;
import java.util.function.Supplier;

public class CommandProcessorTests {
    CommandProcessor processor;
    FakeOutputAdapter output;

    @BeforeEach
    public void setup(){
        output = new FakeOutputAdapter();
        processor = new CommandProcessor();
    }

    @Test
    public void greetCommandShouldGreetMe(){
        processor.addCommand("greet", new GreetCommand("jean", "dupont"));
        processor.processCommand("greet", output::sendOut);
        Assertions.assertEquals("Hello, jean dupont!", output.getAllOutputs());
    }

    @Test
    public void exitCommandShouldSayGoodbye(){
        processor.addCommand("exit", new ExitCommand());
        processor.processCommand("exit", output::sendOut);
        Assertions.assertEquals("Goodbye!", output.getAllOutputs());
    }

    @Test
    public void helpCommandShouldShowAllCommandsAvailable(){
        String expected = """
                Here are all the available commands:
                exit
                help""";
        processor.addCommand("exit", new ExitCommand());
        processor.addCommand("help", new HelpCommand(processor.getCommandsNames()));
        processor.processCommand("help", output::sendOut);
        Assertions.assertEquals(expected, output.getAllOutputs());
    }

    @Test
    public void birthdayCommandShouldTellYouHowManyDaysLeftBeforeYourBirthday(){
        Supplier<LocalDate> todaySupplier = () -> LocalDate.of(2024, 8, 4); ;  // Supplier using method reference
        LocalDate birthday = LocalDate.of(2024, 8, 14);  // Example birthday

        processor.addCommand("birthday", new BirthdayCommand(todaySupplier, birthday));
        processor.processCommand("birthday", output::sendOut);
        Assertions.assertEquals("Still 10 days left!", output.getAllOutputs());
    }
}

File: solution/kotlin/week05-exercise2/src/test/kotlin/CommandProcessorTests.kt
package command

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import java.time.LocalDate
import java.util.function.Supplier

class CommandProcessorTests : StringSpec({
    lateinit var processor: CommandProcessor
    lateinit var output: FakeOutputAdapter

    beforeEach {
        output = FakeOutputAdapter()
        processor = CommandProcessor()
    }

    "greetCommandShouldGreetMe" {
        processor.addCommand("greet", GreetCommand("jean", "dupont"))
        processor.processCommand("greet", output::sendOut)
        output.getAllOutputs() shouldBe "Hello, jean dupont!"
    }

    "exitCommandShouldSayGoodbye" {
        processor.addCommand("exit", ExitCommand())
        processor.processCommand("exit", output::sendOut)
        output.getAllOutputs() shouldBe "Goodbye!"
    }

    "helpCommandShouldShowAllCommandsAvailable" {
        val expected = """
            Here are all the available commands:
            exit
            help
        """.trimIndent()
        processor.addCommand("exit", ExitCommand())
        processor.addCommand("help", HelpCommand(processor.commandsNames))
        processor.processCommand("help", output::sendOut)
        output.getAllOutputs() shouldBe expected
    }

    "birthdayCommandShouldTellYouHowManyDaysLeftBeforeYourBirthday" {
        val todaySupplier: Supplier<LocalDate> = Supplier { LocalDate.of(2024, 8, 4) }
        val birthday = LocalDate.of(2024, 8, 14)

        processor.addCommand("birthday", BirthdayCommand(todaySupplier, birthday))
        processor.processCommand("birthday", output::sendOut)
        output.getAllOutputs() shouldBe "Still 10 days left!"
    }
})
